### DISCUSSION TITLE
How to give a client the possibility to get the list of users?

### DISCUSSION THREAD

**Discussion by flixman** [keycloak/keycloak#29372] [Q&A]
I have a client with a service role, and I have given it all the roles of the realm-management client. Still, when I execute this curl
```
curl -sk https://keycloak.local:8443/admin/realms/<realm>/users -d grant_type=client_credentials -d client_id=<client_id> -d client_secret=<client_secret>
```
I get "Unauthorized access". What am I doing wrong?

**Comment by codespearhead** [comment]
I'm not sure I understand your question.

Are you struggling to authenticate against the [Keycloak Admin REST API](https://www.keycloak.org/docs-api/22.0.1/rest-api/index.html)?

If so, see [this answer](https://github.com/keycloak/keycloak/discussions/29090#discussioncomment-9230678) from #29090 .

**Reply by flixman** [reply]
What I mean is: For my application I am creating a public client for the frontend, a confidential client for the backend, and an additional test client for the tests. I need that test client to be able to retrieve the list of users and, to that end, I created it as a service role and given all the roles from realm-management.... but when I try to get the list of users using the previously shared command, it returns "unauthorized".

**Reply by codespearhead** [reply]
Does the suggested script in that discussion answer your question?

**Reply by flixman** [reply]
no, because I can successfully authenticate using the admin-cli client. What I need is to have a test client with this specific permission to query users, and I do not know what roles it requires.

**Reply by codespearhead** [reply]
Try assigning the `view-users` role to that `confidential` client (Keycloak client with `client authentication` and `authorization` fields turned on) of yours.

![dicussion-29372-1](https://github.com/keycloak/keycloak/assets/72931357/1b612d72-8f5c-4c7f-a9c8-457d1011f27b)

**Reply by flixman** [reply]
This is what I have, already:
![image](https://github.com/keycloak/keycloak/assets/11373728/15296e3e-9530-4fb7-a3ad-3bdf50f48f4c)
It's because of this, that I do not know what is wrong.

**Reply by flixman** [reply]
I do not understand what I am doing wrong with curl. However, with python (using authlib, OAuth2Client) I can use that test client as expected (after fetching the token and so). Thank you very much for your help, anyway, @codespearhead !
